src/MessageWriter.js

```diff
@@ -0,0 +1,312 @@
+        }
+
+        // start the for-loop
+        writerLines.push(`for (var i = 0; i < ${lenField}; i++) {`);
+        // if the sub type is complex we need to allocate it and parse its values
+        if (def.isComplex) {
+          const defType = findTypeByName(types, def.type);
+          // recursively call the function for the sub-type
+          writerLines.push(`  ${friendlyName(defType.name)}(${argName}, ${accessMessageField}[i]);`);
+        } else {
+          // if the subtype is not complex its a simple low-level operation
+          writerLines.push(`  ${argName}.${def.type}(${accessMessageField}[i]);`);
+        }
+        writerLines.push("}"); // close the for-loop
+      } else if (def.isComplex) {
+        const defType = findTypeByName(types, def.type);
+        writerLines.push(`${friendlyName(defType.name)}(${argName}, ${accessMessageField});`);
+      } else {
+        // Call primitives directly.
+        writerLines.push(`${argName}.${def.type}(${accessMessageField});`);
+      }
+    });
+    return writerLines.join("\n    ");
+  };
+
+  let writerJs = "";
+  let calculateSizeJs = "";
+
+  namedTypes.forEach((t) => {
+    writerJs += `
+  function ${friendlyName(t.name)}(writer, message) {
+    ${constructorBody(t, "writer")}
+  };\n`;
+    calculateSizeJs += `
+  function ${friendlyName(t.name)}(offsetCalculator, message) {
+    ${constructorBody(t, "offsetCalculator")}
+  };\n`;
+  });
+
+  writerJs += `
+  return function write(writer, message) {
+    ${constructorBody(unnamedType, "writer")}
+    return writer.buffer;
+  };`;
+  calculateSizeJs += `
+  return function calculateSize(offsetCalculator, message) {
+    ${constructorBody(unnamedType, "offsetCalculator")}
+    return offsetCalculator.offset;
+  };`;
+
+  let _write: (writer: StandardTypeWriter, message: any) => Buffer;
+  let _calculateSize: (offsetCalculator: StandardTypeOffsetCalculator, message: any) => number;
+  try {
+    _write = eval(`(function buildWriter() { ${writerJs} })()`);
+  } catch (e) {
+    console.error("error building writer:", writerJs); // eslint-disable-line
+    throw e;
+  }
+  try {
+    _calculateSize = eval(`(function buildSizeCalculator() { ${calculateSizeJs} })()`);
+  } catch (e) {
+    console.error("error building size calculator:", calculateSizeJs); // eslint-disable-line
+    throw e;
+  }
+
+  return function(message: any) {
+    const offsetCalculator = new StandardTypeOffsetCalculator();
+    const bufferSize = _calculateSize(offsetCalculator, message);
+    const buffer = new Buffer(bufferSize);
+    const writer = new StandardTypeWriter(buffer);
+    return _write(writer, message);
+  };
+}
+
+export class MessageWriter {
+  writer: (message: any) => Buffer;
+
+  // takes a multi-line string message definition and returns
+  // a message reader which can be used to write messages based
+  // on the message definition
+  constructor(messageDefinition: string) {
```

Comment:
weren't we going to use `RosMsgDefinition[]` here instead of the stringified def?

Reply:
I decided to keep it identical to MessageReader for this PR

Reply:
just saw everything is now `RosMsgDefinition` -- why the change of heart?

Reply:
@troygibb I decided to fold it in with this one instead of making a separate PR, and it makes more sense when I started messing around with manually creating message definitions.